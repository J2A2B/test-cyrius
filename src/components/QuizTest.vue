<template>
  <div class="entry-quiz">
    <h1 id="user-text">{{ titleQuiz }}</h1>
    <div v-if="!displayTextEnd">
      <template v-for="(question, i) in data.questions">
        <AskQuestion
          v-if="i === data.questionIndex"
          :key="i"
          :index="i"
          :question="question"
          @emitAnswer="triggerAnswer"
        />
      </template>
    </div>
    <div v-else>
      <p id="success-message" v-if="isCorrectAnswer" class="success">
        {{ WORDINGPAGE.success }}
      </p>
      <p id="error-message" v-else class="error">
        {{ WORDINGPAGE.error }}
      </p>
    </div>
  </div>
</template>

<script lang="ts" setup>
import { defineProps, reactive, computed } from "vue";
import AskQuestion from "./AskQuestion.vue";

interface Question {
  text: string;
  answer: boolean | null;
  correctAnswer: boolean;
}
interface Value {
  index: number;
  answer: boolean | null;
}
interface WordingPage {
  success: string;
  error: string;
}

const WORDINGPAGE: WordingPage = {
  success: "Bravo tout juste !",
  error: "Raté ! Tu as fait des erreurs",
};
const data = reactive<{ questions: Question[]; questionIndex: number }>({
  questions: [
    {
      text: "La couleur du soleil est jaune",
      answer: null,
      correctAnswer: true,
    },
    {
      text: "La couleur du ciel est rouge pale",
      answer: null,
      correctAnswer: false,
    },
    {
      text: "Le président Françasi s'appelle Marc",
      answer: null,
      correctAnswer: false,
    },
  ],
  questionIndex: 0,
});
defineProps({
  titleQuiz: { type: String, default: "" },
});
const triggerAnswer = (value: Value) => {
  data.questions[value.index].answer = value.answer;
  data.questionIndex++;
};
const displayTextEnd = computed<boolean>(() =>
  data.questions.every((el) => el.answer !== null)
);
const isCorrectAnswer = computed<boolean>(() =>
  data.questions.every((el) => el.answer === el.correctAnswer)
);
</script>

<style>
.entry-quiz {
  border-radius: 5px;
  border: 1px solid grey;
  width: fit-content;
  margin: auto;
  padding-bottom: 20px;
  padding-right: 50px;
  padding-left: 50px;
}
.success {
  border-radius: 5px;
  padding: 5px;
  background-color: green;
  color: white;
}
.error {
  border-radius: 5px;
  padding: 5px;
  background-color: red;
  color: white;
}
</style>
