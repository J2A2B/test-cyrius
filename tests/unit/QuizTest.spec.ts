import { shallowMount, mount } from "@vue/test-utils";
import QuizTest from "@/components/QuizTest.vue";

describe("QuizTest.vue", () => {
  describe("Template", () => {
    it("should match snapshot", async () => {
      const wrapper = shallowMount(QuizTest);

      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe("Data", () => {
    it("should set the default Data values", async () => {
      const wrapper = mount(QuizTest);

      expect(wrapper.vm.data.questionIndex).toEqual(0);
      expect(wrapper.vm.data.questions).toEqual([
        {
          text: "La couleur du soleil est jaune",
          answer: null,
          correctAnswer: true,
        },
        {
          text: "La couleur du ciel est rouge pale",
          answer: null,
          correctAnswer: false,
        },
        {
          text: "Le président Françasi s'appelle Marc",
          answer: null,
          correctAnswer: false,
        },
      ]);
    });
    describe("When all Datas are filled wrong", () => {
      it("should display the error message", async () => {
        const wrapper = mount(QuizTest);
        const expectedQuestions = [
          {
            text: "La couleur du soleil est jaune",
            answer: false,
            correctAnswer: true,
          },
          {
            text: "La couleur du ciel est rouge pale",
            answer: true,
            correctAnswer: false,
          },
          {
            text: "Le président Françasi s'appelle Marc",
            answer: false,
            correctAnswer: false,
          },
        ];
        wrapper.vm.data.questions = expectedQuestions;
        await wrapper.vm.$nextTick();
        expect(wrapper.find("p#error-message").text()).toEqual(
          "Raté ! Tu as fait des erreurs"
        );
      });
    });
    describe("When all Datas are filled good", () => {
      it("should display the success message", async () => {
        const wrapper = mount(QuizTest);
        const expectedQuestions = [
          {
            text: "La couleur du soleil est jaune",
            answer: true,
            correctAnswer: true,
          },
          {
            text: "La couleur du ciel est rouge pale",
            answer: false,
            correctAnswer: false,
          },
          {
            text: "Le président Françasi s'appelle Marc",
            answer: false,
            correctAnswer: false,
          },
        ];
        wrapper.vm.data.questions = expectedQuestions;
        await wrapper.vm.$nextTick();
        expect(wrapper.find("p#success-message").text()).toEqual(
          "Bravo tout juste !"
        );
      });
    });
  });
  describe("Props", () => {
    describe("When the props titleQuiz is filled", () => {
      it("should display the title of the quiz", async () => {
        const titleQuiz = "My title";
        const wrapper = shallowMount(QuizTest, {
          props: {
            titleQuiz,
          },
        });
        expect(wrapper.find("h1#user-text").text()).toEqual(titleQuiz);
      });
    });
  });
});
