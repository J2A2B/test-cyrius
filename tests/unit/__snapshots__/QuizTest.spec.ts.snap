// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizTest.vue Template should match snapshot 1`] = `
<div
  class="entry-quiz"
>
  <h1
    id="user-text"
  />
  <div>
    
    
    <ask-question-stub
      index="0"
      question="[object Object]"
    />
    
    
    <!--v-if-->
    
    
    <!--v-if-->
    
    
  </div>
</div>
`;
